### PR TITLE
change binary path and service unit file permission

### PR DESCRIPTION
**Pull request by nagworld9**
Failed to run test service
Dec 27 23:24:51 red8 systemd[6842]: gatestext.service: Failed to execute command: Permission denied
Dec 27 23:24:51 red8 systemd[6842]: gatestext.service: Failed at step EXEC spawning /var/lib/waagent/Microsoft.Azure.Extensions.Edp.GATestExtGo-in/gatestext_script_linux: Permission denied

### CODE REVIEW

File: scripts/service_install.sh [hunk 1]
```diff
@@ -4,6 +4,7 @@ set -euxo pipefail
 service_name="gatestext.service"
 
 dir=$(pwd)
+cp $dir/bin/gatestext_script_linux /usr/local/bin
 eval "sed s#{WORKDIR}#$dir# $dir/services/gatestext.service > /lib/systemd/system/gatestext.service"
-chmod 755 /lib/systemd/system/$service_name
-systemctl daemon-reload
\ No newline at end of file
+chmod 644 /lib/systemd/system/$service_name
```
Comment (larohra) [inline scripts/service_install.sh]: Did you verify if these new permissions work in Ubuntu?

File: scripts/service_install.sh [hunk 2]
```diff
@@ -4,6 +4,7 @@ set -euxo pipefail
 service_name="gatestext.service"
 
 dir=$(pwd)
+cp $dir/bin/gatestext_script_linux /usr/local/bin
```
Comment (larohra) [inline scripts/service_install.sh:10]: Is `/usr/local/bin` common for all distros? Or is it distro dependent? 
Do you need to add any other permissions for this binary? (i.e. making it executable, accessible by root vs non-root, etc?)

Please verify all that by running it in all the distro's before merging it in

File: scripts/service_install.sh (hunk 1 above)
Comment (larohra) [reply scripts/service_install.sh]: As per [this doc on redhat systemd files](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/system_administrators_guide/chap-managing_services_with_systemd), the location might need to be changed - 

<html>
<body>
<p class="title" style="box-sizing: border-box; margin: 0px 0px 1.8em; line-height: 1.5em; color: rgb(37, 37, 37); font-family: &quot;Red Hat Text&quot;, RedHatText, &quot;Helvetica Neue&quot;, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><strong style="box-sizing: border-box; font-weight: 700;">Table 10.2. Systemd Unit Files Locations</strong></p><div class="table-contents" style="box-sizing: border-box; overflow-x: auto; width: 50vw; max-width: 75rem; color: rgb(37, 37, 37); font-family: &quot;Red Hat Text&quot;, RedHatText, &quot;Helvetica Neue&quot;, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Directory | Description
-- | --
/usr/lib/systemd/system/ | Systemd unit files distributed with installed RPM packages.
/run/systemd/system/ | Systemd unit files created at run time. This directory takes precedence over the directory with installed service unit files.
/etc/systemd/system/ | Systemd unit files created by systemctl enable as well as unit files added for extending a service. This directory takes precedence over the directory with runtime unit files.

</div>
</body>
</html>

Comment (larohra) [reply scripts/service_install.sh]: For ubuntu, the path you provided seems correct, but for redhat its different. 

From [this post](https://askubuntu.com/questions/876733/where-are-the-systemd-units-services-located-in-ubuntu) - 

> The package-provided service files are all usually located in /lib/systemd/system. For example, search for .service in the package index.
From man systemd.unit:
/etc/systemd/system/*
/run/systemd/system/*
/lib/systemd/system/*
...

Given the discrepancies between multiple distros, please devise a way to make this path distro dependent vs just a static path. Thanks!

Comment (nagworld9) [reply scripts/service_install.sh]: I verified for ubuntu and did work. At this point enabled distros still works with /lib/systemd/system static path since redhat distros have symlink `/lib is a symlink to /usr/lib`.

File: scripts/service_install.sh (hunk 2 above)
Comment (nagworld9) [reply scripts/service_install.sh:10]: Did work for ubuntu, redhat and centos without changing any sort of permissions by placing it in` /usr/local/bin` path

File: main/systemd.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        for _, v := range distros {
+            if v == distro {
+                return true
+            }
+        }
+    }
+    return false
+}
+
+func get_distro() string {
```
Comment (larohra) [inline main/systemd.go]: NIT :Function names in golang follow `camelCasing` over `snake_casing`. Some normal conventions - https://www.golangprograms.com/naming-conventions-for-golang-functions.html

File: scripts/distro_script.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+set -euo pipefail
+
+distro=None
+if [ -f /etc/os-release ]; then
```
Comment (larohra) [inline scripts/distro_script.sh:5]: All this script is doing is checking if certain files exist and based on that decide the distro. Why not just do this in golang itself? Why add this extra layer?

Comment (larohra) [reply scripts/distro_script.sh:5]: ohh nvm I see some extra logic where we're executing some files based on some conditions. I'm not a big fan of doing external calls but since its just a test extension I'm ok with this.

File: main/systemd.go (hunk 3 above)
Comment (nagworld9) [reply main/systemd.go]: Thanks for pointing out. I thought I followed everywhere but somehow, I missed this

File: scripts/service_install.sh [hunk 5]
```diff
@@ -4,6 +4,7 @@ set -euxo pipefail
 service_name="gatestext.service"
 
 dir=$(pwd)
-eval "sed s#{WORKDIR}#$dir# $dir/services/gatestext.service > /lib/systemd/system/gatestext.service"
-chmod 755 /lib/systemd/system/$service_name
-systemctl daemon-reload
\ No newline at end of file
+cp $dir/bin/gatestext_script_linux /usr/local/bin
+cp $dir/services/gatestext.service $1
```
Comment (larohra) [inline scripts/service_install.sh:11]: Can you add a comment about what the expectation for the `$1` path is

File: main/systemd.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    distroName := strings.ToLower(string(output))
+    infoLogger.Println(fmt.Sprintf("distro name: %s", distroName))
+    return distroName
+}
+
+func getSystemdUnitFileInstallPath() string {
+
+    var systemdPath string
+
+    distro := get_distro()
```
Comment (larohra) [inline main/systemd.go]: what if this function faces an issue? you're not passing the errors as expected for golang. Please add that

File: main/systemd.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    if err != nil {
+        errorMessage := fmt.Sprintf("Error getting distro name: %+v", err)
+		errorLogger.Println(errorMessage)
+    }
+    distroName := strings.ToLower(string(output))
+    infoLogger.Println(fmt.Sprintf("distro name: %s", distroName))
+    return distroName
+}
+
+func getSystemdUnitFileInstallPath() string {
```
Comment (larohra) [inline main/systemd.go]: NIT: Since this function is being used outside this module, shouldn't this start with a capital letter? 
```suggestion
func GetSystemdUnitFileInstallPath() string {
```

Comment (nagworld9) [reply main/systemd.go]: `If the functions with names that start with an uppercase letter will be exported to other packages. If the function name starts with a lowercase letter, it won't be exported to other packages, but you can call this function within the same package.`

Our case this function is being used in same package

File: main/systemd.go (hunk 6 above)
Comment (nagworld9) [reply main/systemd.go]: what sort of errors we expect from get_distro. Already checking inside function and logging appropriately.

File: main/systemd.go (hunk 7 above)
Comment (larohra) [reply main/systemd.go]: I mean for now yes, but since its a utility function it can be imported too dont you think?

File: main/systemd.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+func get_distro() string {
+    path, _ := os.Getwd()
+    output, err := exec.Command("sudo", filepath.Join(path, "scripts/distro_script.sh")).Output()
+    if err != nil {
+        errorMessage := fmt.Sprintf("Error getting distro name: %+v", err)
+		errorLogger.Println(errorMessage)
+    }
+    distroName := strings.ToLower(string(output))
```
Comment (larohra) [inline main/systemd.go:29]: If the previous call fails, the value of `output` would be incorrect, we should not try to derive the distro name in that case and fail the function

File: main/systemd.go (hunk 6 above)
Comment (larohra) [reply main/systemd.go]: > what sort of errors we expect from get_distro

Right out of the bat, the first thing that comes to mind is what if your shell script fails with a non-0 exit code? The script is using a lot of different utilities based on some files but what if something is wrong there and it fails? What if we don;t have sudo privileges to run this command?

Plethora of things can go wrong, there's no harm in having a proper error handling in place. 

For your use case though, I don't mind returning the default `systemd`  path value incase we're unable to determine the distro, but that check should be inside this function rather than the `getDistro` function, the `getDistro` should still return an error incase it can't determine the distro. 

Here's a useful link I found about error handling in golang, might be a worthwhile read - https://gabrieltanner.org/blog/golang-error-handling-definitive-guide

File: main/systemd.go (hunk 8 above)
Comment (nagworld9) [reply main/systemd.go:29]: I tested it, you get empty output and won't break the function. At the end of it we get default systemdpath if distro not found.

File: main/systemd.go (hunk 6 above)
Comment (nagworld9) [reply main/systemd.go]: `Right out of the bat, the first thing that comes to mind is what if your shell script fails with a non-0 exit code? The script is using a lot of different utilities based on some files but what if something is wrong there and it fails? What if we don;t have sudo privileges to run this command?`

These errors you get from from stderr, you pass this to and catch at same place doesn't make a difference in this scenario. we still achieving same goal. I don't mind changing and following convention.

Comment (larohra) [reply main/systemd.go]: > I don't mind changing and following convention

Functionally it doesn't make sense, but at the same time convention is what we need to keep the state of the repo clean, easy to read and make it futureproof. I would really appreciate it if you did follow it, thank you :)

File: main/systemd.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+            }
+        }
+    }
+    return false
+}
+
+func get_distro() string {
+    path, _ := os.Getwd()
+    output, err := exec.Command("sudo", filepath.Join(path, "scripts/distro_script.sh")).Output()
+    if err != nil {
```
Comment (larohra) [inline main/systemd.go:26]: Is this condition true `err != nil` if the exit code is non-0? If so, what's the error? Does it just contain the non-0 exit code?

File: main/systemd.go (hunk 8 above)
Comment (larohra) [reply main/systemd.go:29]: wrong place for handling this functionality. This default behavior should be one level up, not in this utility function

File: main/systemd.go (hunk 9 above)
Comment (nagworld9) [reply main/systemd.go:26]: Yes, it does. I see error codes (1, 123, etc..) logging on failure cases

Comment (larohra) [reply main/systemd.go:26]: cool thanks

File: scripts/service_install.sh (hunk 5 above)
Comment (nagworld9) [reply scripts/service_install.sh:11]: addressed

File: main/systemd.go (hunk 8 above)
Comment (nagworld9) [reply main/systemd.go:29]: Actually, default behavior was in another function not here. I have changed now as part of error handling.

File: main/systemd.go (hunk 6 above)
Comment (nagworld9) [reply main/systemd.go]: included error handling

File: main/systemd.go (hunk 8 above)
Comment (kevinclark19a) [reply main/systemd.go:29]: Would `err` not be non-nil in the cause of the previous call failing? I think this is covered by the early return.

File: main/systemd.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+func GetSystemdUnitFileInstallPath() string {
+
+    var systemdPath string
+
+    distro, err := getDistro()
+
+    if (err != nil) {
+        errorMessage := fmt.Sprintf("Error getting distro name: %+v : %s", err, err.Error())
+	    errorLogger.Println(errorMessage)
```
Comment (kevinclark19a) [inline main/systemd.go]: NIT: This looks like a formatting issue. Can you fix it?

File: main/systemd.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    } else {
+        infoLogger.Println(fmt.Sprintf("distro name: %s", distro))
+    }
+
+    distros1 :=[]string{"ubuntu", "debian"}
+    distros2 :=[]string{"suse", "sle_hpc", "sles", "opensuse", "redhat", "rhel", "centos", "oracle"}
```
Comment (kevinclark19a) [inline main/systemd.go]: Could you use more descriptive names for these lists?

Comment (larohra) [reply main/systemd.go]: +1

**Review by larohra — DISMISSED** [review]

File: main/systemd.go (hunk 8 above)
Comment (nagworld9) [reply main/systemd.go:29]: yeah, err would be non-nil in case of failure and returning error which I have addressed now as part of error handling

File: main/systemd.go (hunk 10 above)
Comment (nagworld9) [reply main/systemd.go]: I noticed this. It's funny my IDE showing inline but GitHub shows formatting off.

File: main/systemd.go (hunk 11 above)
Comment (nagworld9) [reply main/systemd.go]: updated

File: main/systemd.go (hunk 10 above)
Comment (nagworld9) [reply main/systemd.go]: had to move this statement back and forth to figure out formatting issue

File: main/systemd.go [hunk 12]
```diff
@@ -38,17 +38,17 @@ func GetSystemdUnitFileInstallPath() string {
...
-	    errorLogger.Println(errorMessage)
+        errorLogger.Println(errorMessage)
     } else {
         infoLogger.Println(fmt.Sprintf("distro name: %s", distro))
     }
 
-    distros1 :=[]string{"ubuntu", "debian"}
-    distros2 :=[]string{"suse", "sle_hpc", "sles", "opensuse", "redhat", "rhel", "centos", "oracle"}
+    identicalSystemdPathDistros1 :=[]string{"ubuntu", "debian"}
+    identicalSystemdPathDistros2 :=[]string{"suse", "sle_hpc", "sles", "opensuse", "redhat", "rhel", "centos", "oracle"}
```
Comment (kevinclark19a) [inline main/systemd.go]: I'm not sure these are much better... I was thinking more `debianDistros` vs `redhatDistros`.

Comment (nagworld9) [reply main/systemd.go]: Honestly this list mix of all flavors, it's hard to come up with standard name. That's why I don't want to have specific name. I don't think 100% convinced with proposed names but if it looks better in overall sense than others, I'm good to change.

Comment (nagworld9) [reply main/systemd.go]: @kevinclark19a Changed to proposed names

**Review by kevinclark19a — DISMISSED** [review]

File: main/main.go [hunk 13]
```diff
@@ -21,7 +21,7 @@ import (
 var (
 	versionMajor    = "1"
 	versionMinor    = "1"
-	versionBuild    = "0"
+	versionBuild    = "1"
```
Comment (larohra) [inline main/main.go]: Version update should be a separate PR

**Review by larohra — CHANGES_REQUESTED** [review]
Please revert the version update, that should be a separate PR

File: main/main.go (hunk 13 above)
Comment (nagworld9) [reply main/main.go]: Last time, I have received comment to change it in same PR. So, I included this time.

Comment (larohra) [reply main/main.go]: Sorry about that, I was wrong that time :)

Comment (nagworld9) [reply main/main.go]: No worries, I will undo and open separate PR.

**Review by larohra — APPROVED** [review]

**Review by kevinclark19a — APPROVED** [review]
